backend/server.js

```diff
@@ -2,28 +2,152 @@ import express from 'express'
+  // If no arguments, returns true if any path in this document is modified."
+  // https://mongoosejs.com/docs/api.html#document_Document-isModified
+  if (!user.isModified('password')) {
+    return next()
+  }
+  const salt = bcrypt.genSaltSync(10)
+  // Hash the password – this happens after the validation.
+  user.password = bcrypt.hashSync(user.password, salt)
+  next()
+})
+
+//_________Model
+const User = mongoose.model('User', userSchema)
+
+
+//_________Defines port 
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
+//_________Add middlewares to enable cors and json body parsing
 app.use(cors())
 app.use(bodyParser.json())
 
-// Start defining your routes here
+//_________Error message if server is down
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).send({ error: SERVICE_UNAVAILABLE })
+  }
+})
+
+//_________middlewear to authenticate User
+const authenticateUser = async (req, res, next) => {
+  try {
+    const user = await User.findOne({
+      accessToken: req.header('Authorization'),
+    })
+
+    if (user) {
+      req.user = user
+      next()
+    } else {
+      res
+        .status(401)
+        .json({ loggedOut: true, message: USER_NOT_FOUND })
+    }
+  } catch (err) {
+    res
+      .status(403)
+      .json({ message: LOGIN_FAILED, errors: err })
+  }
+}
+
+//_________GET list of endpoints
+const listEndpoints = require('express-list-endpoints')
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+//_________POST create user
+app.post('/signup', async (req, res) => {
+  try {
+    const { name, email, password } = req.body
+    const user = await new User({ name, email, password }).save()
+    res.status(201).json({ userId: user._id, userName: user.name, accessToken: user.accessToken })
+  } catch (err) {
+    res.status(400).json({
+      message: SING_UP_FAILED, errors: {
+        message: err.message,
+        error: err,
+      },
+    })
+  }
+})
+
+//_________Secure endpoint, user needs to be logged in to access this.
+app.get('/users/:id/secret', authenticateUser)
+
+//_________POST secretMessage
+app.get('/users/:id/secret', (req, res) => {
+  const secretMessage = `${req.user.name}, this was totally worth all your effort - right?`
+  res.status(201).json(secretMessage)
+})
+
+//_________POST Log in user endpoint (POST)
+app.post('/login', async (req, res) => {
+  try {
+    const { name, email, password } = req.body
+    const user = await User.findOne({ name, email })
+    if (user && bcrypt.compareSync(password, user.password)) {
+      res.status(201).json({ userId: user._id, userName: user.name, accessToken: user.accessToken })
+    } else {
+      res.status(404).json({ notFound: true })
+    }
+  } catch (err) {
+    res.status(404).json({ notFound: true })
```

Comment:
Maybe throw an error in the else here and also for consistency use consts for error-messages. Not found: true maybe doesn't give enough info to the user? Perhaps have a more specific message.